### PR TITLE
Fix unchain gradient pull

### PR DESCRIPTION
**Pull request by Rishav1**
This change would use Null gradients for child nodes that have been
unchained. References issue #6027. As opposed to current behavior:

```python
>>> import numpy as np; import chainer; import chainer.functions as F
>>> x = chainer.Variable(np.arange(4).astype('f').reshape(2, 2)); x
variable([[0., 1.],
          [2., 3.]])
>>> h0, h1 = F.split_axis(x, [1], axis=0); h0, h1
(variable([[0., 1.]]), variable([[2., 3.]]))
>>> y = F.sum(h0)
>>> z = F.sum(h1)
>>> h0.unchain()
>>> x.grad, h0.grad, h1.grad
(None, None, None)
>>> y.backward()
>>> x.grad, h0.grad, h1.grad
(None, array([[1., 1.]], dtype=float32), None)
>>> z.backward()
>>> x.grad, h0.grad, h1.grad
(array([[1., 1.],
       [1., 1.]], dtype=float32), None, None)
```
we would get this:

```python
>>> import numpy as np; import chainer; import chainer.functions as F
>>> x = chainer.Variable(np.arange(4).astype('f').reshape(2, 2)); x
variable([[0., 1.],
          [2., 3.]])
>>> h0, h1 = F.split_axis(x, [1], axis=0); h0, h1
(variable([[0., 1.]]), variable([[2., 3.]]))
>>> y = F.sum(h0)
>>> z = F.sum(h1)
>>> h0.unchain()
>>> x.grad, h0.grad, h1.grad
(None, None, None)
>>> y.backward()
>>> x.grad, h0.grad, h1.grad
(None, array([[1., 1.]], dtype=float32), None)
>>> z.backward()
>>> x.grad, h0.grad, h1.grad
(array([[0., 0.],
       [1., 1.]], dtype=float32), array([[1., 1.]], dtype=float32), None)
```

Thank you for creating a pull request!

Please double-check the following.

- Read [our contribution guide](https://docs.chainer.org/en/stable/contribution.html).
  - Does your code conform to our coding guidelines?
  - Did you write sufficient test code?
  - Did you write sufficient documentation?
- Also, take a look at [our compatibility policy](https://docs.chainer.org/en/stable/compatibility.html).

### CODE REVIEW

File: chainer/variable.py [hunk 1]
```diff
@@ -1614,7 +1614,8 @@ def add_cand(cand):
             i for i, x in enumerate(inputs) if x.requires_grad
         ])
         outputs = [y() for y in func.outputs]  # access via weak ref
-        out_grad = tuple([grads.pop(y) for y in outputs])
+        out_grad = tuple([grads.pop(y) if y.creator_node else None
```
Comment (toslunar) [inline chainer/variable.py]: - `y` can be `None` if the output (node) is deleted.
- `func is None` is better because `func.__bool__` might be overridden.
```suggestion
        out_grad = tuple([grads.pop(y)
                          if y is not None and y.creator_node is not None
                          else None
```

File: tests/chainer_tests/test_function_node.py [hunk 2]
```diff
@@ -869,6 +869,29 @@ def test_length_check(self):
             chainer.grad([y], [x], [None], [None, None])
 
 
+class TestUnchainSplit(unittest.TestCase):
```
Comment (toslunar) [inline tests/chainer_tests/test_function_node.py]: Since the test is somewhat corner-case, I'd place this test after the basic cases.  Could you move it to just after `TestGradComplex` or just before `TestGradV3Compat1`?

Could you rename the class?
```suggestion
class TestGradUnchainSplit(unittest.TestCase):
```

File: tests/chainer_tests/test_function_node.py [hunk 3]
```diff
@@ -869,6 +869,29 @@ def test_length_check(self):
...
+        w = y + z
+        h0.unchain()
+
+        dy_dh0 = numpy.array([[1., 1.]])
+        dz_dh1 = numpy.array([[1., 1.]])
+        dy_dx = None
+        dz_dx = numpy.array([[0., 0.], [1., 1.]])
+        dw_dx = numpy.array([[0., 0.], [1., 1.]])
+
+        assert numpy.all(chainer.grad([y], [h0])[0].array == dy_dh0)
```
Comment (toslunar) [inline tests/chainer_tests/test_function_node.py]: Could you use `testing.assert_allclose`, to be ready for adding GPU tests in future?

File: tests/chainer_tests/test_variable.py [hunk 4]
```diff
@@ -507,6 +507,17 @@ def test_unchain(self):
         assert ret[1].rank == old_rank
         self.check_backward((ret[1],), (ret[2],), (ret[3],), False)
 
+    def test_unchain_split(self):
+        if not len(self.x.shape) > 0:
```
Comment (toslunar) [inline tests/chainer_tests/test_variable.py]:
```suggestion
        if self.x.ndim == 0:
```

File: tests/chainer_tests/test_variable.py [hunk 5]
```diff
@@ -507,6 +507,17 @@ def test_unchain(self):
...
+            return
+        ret = get_variable(np, self.x)
+        ret.grad = np.zeros_like(ret.data)
+        y1, y2 = F.split_axis(ret, [5], axis=0)
+        y1.unchain()
+        z1, z2 = F.sum(y1), F.sum(y2)
+        z1.backward()
+        self.check_backward((ret,), (y1, y2,), (z2,), False)
```
Comment (toslunar) [inline tests/chainer_tests/test_variable.py]: To test `y1.unchain()`, it's more natural to test that `y1` become an "input".  While it'll be less realistic, I recommend something like
```
w = z1 + z2
self.check_backward((ret, y1), (y2, z1, z2), (w,), False)
```

File: tests/chainer_tests/test_variable.py [hunk 6]
```diff
@@ -507,6 +507,17 @@ def test_unchain(self):
         assert ret[1].rank == old_rank
         self.check_backward((ret[1],), (ret[2],), (ret[3],), False)
 
+    def test_unchain_split(self):
+        if not len(self.x.shape) > 0:
+            return
+        ret = get_variable(np, self.x)
```
Comment (toslunar) [inline tests/chainer_tests/test_variable.py]: It looks ok to use `x` for the variable (without the confusion with `self.x`).

Comment (Rishav1) [reply tests/chainer_tests/test_variable.py]: `x` is the class variable.

File: tests/chainer_tests/test_variable.py (hunk 5 above)
Comment (Rishav1) [reply tests/chainer_tests/test_variable.py]: Okay, this is better.

File: tests/chainer_tests/test_variable.py (hunk 4 above)
Comment (Rishav1) [reply tests/chainer_tests/test_variable.py]: Done.

File: tests/chainer_tests/test_function_node.py (hunk 3 above)
Comment (Rishav1) [reply tests/chainer_tests/test_function_node.py]: Ok.

File: tests/chainer_tests/test_function_node.py (hunk 2 above)
Comment (Rishav1) [reply tests/chainer_tests/test_function_node.py]: Alright.

File: tests/chainer_tests/test_variable.py [hunk 7]
```diff
@@ -507,6 +507,17 @@ def test_unchain(self):
...
 
+    def test_unchain_split(self):
+        if self.x.ndim == 0:
+            return
+        ret = get_variable(np, self.x)
+        ret.grad = np.zeros_like(ret.data)
+        y1, y2 = F.split_axis(ret, [5], axis=0)
+        y1.unchain()
+        z1, z2 = F.sum(y1), F.sum(y2)
+        w = z1 + z2
```
Comment (toslunar) [inline tests/chainer_tests/test_variable.py]: Sorry.  I've just found the backward tests in the class don't rely on default grads.  Could you set grad for each variable, including `w`?

Comment (Rishav1) [reply tests/chainer_tests/test_variable.py]: Okay, I guess I missed it too. Added it.

**Comment by toslunar** [comment]
Jenkins, test this please.
